### ISSUE TITLE
Ambiguous for no_follow_redirects configuration

### ISSUE THREAD

**Issue by nutchalum** [prometheus/blackbox_exporter#243]
### Host operating system: output of `uname -a`
linux
### blackbox_exporter version: output of `blackbox_exporter -version`

prom/blackbox-exporter:v0.10.0
### What is the blackbox.yml module config.
    modules:
      https_2xx:
        prober: http
        timeout: 5s
        http:
          valid_status_codes: [200]
          method: GET
          fail_if_not_ssl: true
          no_follow_redirects: true
### What is the prometheus.yml scrape config.
      - job_name: 'kubernetes-ingresses'

        metrics_path: /probe
        params:
          module: [https_2xx]

        kubernetes_sd_configs:
          - role: ingress

        relabel_configs:
          - source_labels: [__meta_kubernetes_ingress_scheme,__address__,__meta_kubernetes_ingress_path]
            regex: (.+);(.+);(.+)
            replacement: ${1}://${2}${3}
            target_label: __param_target
          - target_label: __address__
            replacement: blackbox-exporter.monitoring-system.svc:9115
          - source_labels: [__param_target]
            target_label: instance
          - action: labelmap
            regex: __meta_kubernetes_ingress_label_(.+)
          - source_labels: [__meta_kubernetes_namespace]
            target_label: kubernetes_namespace
          - source_labels: [__meta_kubernetes_ingress_name]
            target_label: kubernetes_name
### What logging output did you get from adding `&debug=true` to the probe URL?
Didn't try this.
### What did you do that produced an error?
this line of configuration
`no_follow_redirects: true`
### What did you expect to see?
I have an application with root path (/) that redirect to somepath (eg. /example)
When I set `no_follow_redirects: true` and probe at "/" it shouldn't follow the redirect.
That means in prometheus `probe_http_status_code` should be 404

### What did you see instead?
`probe_http_status_code` is 301 (or 302 sometimes)

### A little more experiment
When I set `no_follow_redirects: false` and probe at "/" it should follow the redirect.
That means in prometheus `probe_http_status_code` should be 301 (or 302 sometimes)
However, I got 404 instead

**Comment by brian-brazil** [comment]
Can you share the debug log?
Have you checked what's going on with tcpdump?

**Comment by nutchalum** [comment]
Didn't try tcpdump (I'm sorry about that, I'm too busy)
But here's what the log says (I use port-foward to my localhost)
```
> curl http://localhost:9115/config                                                                                             
modules:
  https_2xx:
    prober: http
    timeout: 5s
    http:
      valid_status_codes:
      - 200
      no_follow_redirects: true
      fail_if_not_ssl: true
      method: GET
```
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https%3A%2F%2Fprometheus-monitor.my.domain.com%2F%3Fdebug%3Dtrue
# HELP content_length Length of http content response
# TYPE content_length gauge
content_length 29
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 2.66e-07
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.035809982
# HELP probe_failed_due_to_regex Indicates if probe failed due to regex
# TYPE probe_failed_due_to_regex gauge
probe_failed_due_to_regex 0
# HELP probe_http_redirects The number of redirects
# TYPE probe_http_redirects gauge
probe_http_redirects 1
# HELP probe_http_ssl Indicates if SSL was used for the final redirect
# TYPE probe_http_ssl gauge
probe_http_ssl 1
# HELP probe_http_status_code Response HTTP status code
# TYPE probe_http_status_code gauge
probe_http_status_code 302
# HELP probe_http_version Returns the version of HTTP of the probe response
# TYPE probe_http_version gauge
probe_http_version 1.1
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_ssl_earliest_cert_expiry Returns earliest SSL cert expiry in unixtime
# TYPE probe_ssl_earliest_cert_expiry gauge
probe_ssl_earliest_cert_expiry 1.526342399e+09
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 0
```

Here's is the weird part as I said above.
The config`no_follow_redirects: true` is set to **true**
The gauge `probe_http_redirects` is **1** (which mean it does follow)
and `probe_http_status_code` is **302** (It's redirected)

**Comment by nutchalum** [comment]
More about details

I have deploy prometheus and access it via `IngressResource`
Let's say it's and I'm doing a probing here `example.com/prometheus`
When I type this url in the browser (example.com/prometheus) 

It redirects to example.com/prometheus/targets (which is its endpoint)

```
> curl -L https://example.com/prometheus -vv
*   Trying 172.20.3.94...
* Connected to example.com (172.20.3.94) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: *.thinknet.co.th
* Server certificate: COMODO RSA Domain Validation Secure Server CA
* Server certificate: COMODO RSA Certification Authority
* Server certificate: AddTrust External CA Root
> GET /prometheus HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:14:57 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 47
< Connection: keep-alive
< Location: /prometheus/
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
* Ignoring the response-body
* Connection #0 to host example.com left intact
* Issue another request to this URL: 'https://example.com/prometheus/'
* Found bundle for host example.com: 0x7fd64ae08010
* Re-using existing connection! (#0) with host example.com
* Connected to example.com (172.20.3.94) port 443 (#0)
> GET /prometheus/ HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 302 Found
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:14:57 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 40
< Connection: keep-alive
< Location: /prometheus/graph
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
* Ignoring the response-body
* Connection #0 to host example.com left intact
* Issue another request to this URL: 'https://example.com/prometheus/graph'
* Found bundle for host example.com: 0x7fd64ae08010
* Re-using existing connection! (#0) with host example.com
* Connected to example.com (172.20.3.94) port 443 (#0)
> GET /prometheus/graph HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:14:57 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
```
My point is `no_follow_redirects: true` is somehow working as opposite as it said.
`no_follow_redirects: true` it does follow.
`no_follow_redirects: false` it doesn't follow.

And I'm still confusing, even though it does follow why does it get 301/302 instead 200 (where endpoint lives)

**Comment by nutchalum** [comment]
Here's my `IngressResource` for considering.
I'm using `--web.external-url=https://example.com/prometheus/` in prometheus too.

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: example.com-prometheus
  annotations:
    kubernetes.io/ingress.class: "nginx"
spec:
  tls:
  - hosts:
    - example.com
    secretName: my-secret

  rules:
  - host: example.com
    http:
      paths:
      - path: /prometheus
        backend:
          serviceName: prometheus-federate
          servicePort: 9090
```

**Comment by brian-brazil** [comment]
That output looks fine to me. The result was 302, as that was the result of the final (and only) request we made. As we didn't follow redirects, that's thus a 302/201 status code.

**Comment by nutchalum** [comment]
I changed `no_follow_redirects` to *false* and did the same request
```
> curl http://localhost:9115/config                                                                                             
modules:
  https_2xx:
    prober: http
    timeout: 5s
    http:
      valid_status_codes:
      - 200
      fail_if_not_ssl: true
      method: GET
```

```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https%3A%2F%2Fprometheus-monitor.my.domain.com%2F%3Fdebug%3Dtrue
# HELP content_length Length of http content response
# TYPE content_length gauge
content_length 21
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 2.59e-07
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.04969281
# HELP probe_failed_due_to_regex Indicates if probe failed due to regex
# TYPE probe_failed_due_to_regex gauge
probe_failed_due_to_regex 0
# HELP probe_http_redirects The number of redirects
# TYPE probe_http_redirects gauge
probe_http_redirects 1
# HELP probe_http_ssl Indicates if SSL was used for the final redirect
# TYPE probe_http_ssl gauge
probe_http_ssl 1
# HELP probe_http_status_code Response HTTP status code
# TYPE probe_http_status_code gauge
probe_http_status_code 404
# HELP probe_http_version Returns the version of HTTP of the probe response
# TYPE probe_http_version gauge
probe_http_version 1.1
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_ssl_earliest_cert_expiry Returns earliest SSL cert expiry in unixtime
# TYPE probe_ssl_earliest_cert_expiry gauge
probe_ssl_earliest_cert_expiry 1.526342399e+09
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 0
```
My point is that I want to follow any redirects to the final.
Using curl without follow is 302 (which is right)
Using curl with follow is 200 (which is right too)
However,
When I use exporter and set `no_follow_redirects` to *true*
I got 301/302 (which is right because we didn't follow redirects).
But when I set `no_follow_redirects` to *false*
I got 404 instead (the final should be 2xx, Am I right?)
```
Here's what curl does without follow
> curl https://example.com/prometheus -v 
*   Trying 172.20.3.94...
* Connected to example.com (172.20.3.94) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: *.thinknet.co.th
* Server certificate: COMODO RSA Domain Validation Secure Server CA
* Server certificate: COMODO RSA Certification Authority
* Server certificate: AddTrust External CA Root
> GET /prometheus HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:41:31 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 47
< Connection: keep-alive
< Location: /prometheus/
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
<a href="/prometheus/">Moved Permanently</a>.

* Connection #0 to host example.com left intact
```
Here's what curl does with follow
```
> curl -L https://example.com/prometheus -v
*   Trying 172.20.3.94...
* Connected to example.com (172.20.3.94) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: *.thinknet.co.th
* Server certificate: COMODO RSA Domain Validation Secure Server CA
* Server certificate: COMODO RSA Certification Authority
* Server certificate: AddTrust External CA Root
> GET /prometheus HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:45:24 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 47
< Connection: keep-alive
< Location: /prometheus/
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
* Ignoring the response-body
* Connection #0 to host example.com left intact
* Issue another request to this URL: 'https://example.com/prometheus/'
* Found bundle for host example.com: 0x7f9c4bc10040
* Re-using existing connection! (#0) with host example.com
* Connected to example.com (172.20.3.94) port 443 (#0)
> GET /prometheus/ HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 302 Found
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:45:24 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 40
< Connection: keep-alive
< Location: /prometheus/graph
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
* Ignoring the response-body
* Connection #0 to host example.com left intact
* Issue another request to this URL: 'https://example.com/prometheus/graph'
* Found bundle for host example.com: 0x7f9c4bc10040
* Re-using existing connection! (#0) with host example.com
* Connected to example.com (172.20.3.94) port 443 (#0)
> GET /prometheus/graph HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 07:45:24 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
```

**Comment by brian-brazil** [comment]
The 404 is almost certainly real.

**Comment by nutchalum** [comment]
That's new to me.
Would you mind to explain me why 404 is real? please.

**Comment by brian-brazil** [comment]
I generally presume that the blackbox exporter works correctly. This this is something you need to debug on your end to see why you're returning a 404.

**Comment by nutchalum** [comment]
I've finally be able to trace it.

Here's my manual probe with debug=true
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=example.com%2Fprometheus\&debug\=true              
Logs for the probe:
ts=2017-10-11T08:49:54.203204093Z caller=main.go:116 module=https_2xx target=example.com/prometheus level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-11T08:49:54.203317641Z caller=utils.go:43 module=https_2xx target=example.com/prometheus level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2017-10-11T08:49:54.207155588Z caller=utils.go:50 module=https_2xx target=example.com/prometheus level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address example.com: no suitable address found"
ts=2017-10-11T08:49:54.210243156Z caller=utils.go:63 module=https_2xx target=example.com/prometheus level=info msg="Resolved target address" ip=172.20.3.94
ts=2017-10-11T08:49:54.210277341Z caller=http.go:190 module=https_2xx target=example.com/prometheus level=info msg="Making HTTP request" url=http://[172.20.3.94]/prometheus host=example.com
ts=2017-10-11T08:49:54.212218733Z caller=http.go:150 module=https_2xx target=example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus
ts=2017-10-11T08:49:54.240752371Z caller=http.go:150 module=https_2xx target=example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus/
ts=2017-10-11T08:49:54.266483896Z caller=http.go:150 module=https_2xx target=example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus/graph
ts=2017-10-11T08:49:54.29344738Z caller=http.go:197 module=https_2xx target=example.com/prometheus level=info msg="Received HTTP response" status_code=200
ts=2017-10-11T08:49:54.293566335Z caller=main.go:127 module=https_2xx target=example.com/prometheus level=info msg="Probe succeeded" duration_seconds=0.090314394
```

And here's my debug using prometheus + ingress sd from https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml

```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https%3A%2F%2Fexample.com%2Fprometheus\&debug\=true
Logs for the probe:
ts=2017-10-11T08:48:23.56326671Z caller=main.go:116 module=https_2xx target=https://example.com/prometheus level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-11T08:48:23.563365303Z caller=utils.go:43 module=https_2xx target=https://example.com/prometheus level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2017-10-11T08:48:23.567152995Z caller=utils.go:50 module=https_2xx target=https://example.com/prometheus level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address example.com: no suitable address found"
ts=2017-10-11T08:48:23.570423959Z caller=utils.go:63 module=https_2xx target=https://example.com/prometheus level=info msg="Resolved target address" ip=172.20.3.43
ts=2017-10-11T08:48:23.570459141Z caller=http.go:190 module=https_2xx target=https://example.com/prometheus level=info msg="Making HTTP request" url=https://[172.20.3.43]/prometheus host=example.com
ts=2017-10-11T08:48:23.59473618Z caller=http.go:150 module=https_2xx target=https://example.com/prometheus level=info msg="Received redirect" url=https://[172.20.3.43]/prometheus/
ts=2017-10-11T08:48:23.617006018Z caller=http.go:197 module=https_2xx target=https://example.com/prometheus level=info msg="Received HTTP response" status_code=404
ts=2017-10-11T08:48:23.617050684Z caller=http.go:206 module=https_2xx target=https://example.com/prometheus level=info msg="Invalid HTTP response status code" status_code=404 valid_status_codes=[200]
ts=2017-10-11T08:48:23.617102103Z caller=main.go:129 module=https_2xx target=https://example.com/prometheus level=error msg="Probe failed" duration_seconds=0.053791863
```

The difference is the `target`
where my manual probe doesn't have *scheme* before the url
and ingress sd example has *scheme* before the url

Should we worried about this?

**Comment by brian-brazil** [comment]
It sounds like your https endpoint isn't handling the same things as http.

**Comment by nutchalum** [comment]
What do I probably miss?

Here's probing with **http** scheme
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=http://example.com%2Fprometheus\&debug\=true 
Logs for the probe:
ts=2017-10-11T09:34:43.982501606Z caller=main.go:116 module=https_2xx target=http://example.com/prometheus level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-11T09:34:43.982607761Z caller=utils.go:43 module=https_2xx target=http://example.com/prometheus level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2017-10-11T09:34:43.986021381Z caller=utils.go:50 module=https_2xx target=http://example.com/prometheus level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address example.com: no suitable address found"
ts=2017-10-11T09:34:43.9890656Z caller=utils.go:63 module=https_2xx target=http://example.com/prometheus level=info msg="Resolved target address" ip=172.20.3.94
ts=2017-10-11T09:34:43.989099756Z caller=http.go:190 module=https_2xx target=http://example.com/prometheus level=info msg="Making HTTP request" url=http://[172.20.3.94]/prometheus host=example.com
ts=2017-10-11T09:34:43.991691598Z caller=http.go:150 module=https_2xx target=http://example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus
ts=2017-10-11T09:34:44.019919311Z caller=http.go:150 module=https_2xx target=http://example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus/
ts=2017-10-11T09:34:44.046107403Z caller=http.go:150 module=https_2xx target=http://example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus/graph
ts=2017-10-11T09:34:44.074309274Z caller=http.go:197 module=https_2xx target=http://example.com/prometheus level=info msg="Received HTTP response" status_code=200
ts=2017-10-11T09:34:44.074367673Z caller=main.go:127 module=https_2xx target=http://example.com/prometheus level=info msg="Probe succeeded" duration_seconds=0.091824493
```

Here's probing with **https** scheme
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https://example.com%2Fprometheus\&debug\=true
Logs for the probe:
ts=2017-10-11T09:36:11.419419492Z caller=main.go:116 module=https_2xx target=https://example.com/prometheus level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-11T09:36:11.419527677Z caller=utils.go:43 module=https_2xx target=https://example.com/prometheus level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2017-10-11T09:36:11.422921066Z caller=utils.go:50 module=https_2xx target=https://example.com/prometheus level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address example.com: no suitable address found"
ts=2017-10-11T09:36:11.426296528Z caller=utils.go:63 module=https_2xx target=https://example.com/prometheus level=info msg="Resolved target address" ip=172.20.3.94
ts=2017-10-11T09:36:11.426330606Z caller=http.go:190 module=https_2xx target=https://example.com/prometheus level=info msg="Making HTTP request" url=https://[172.20.3.94]/prometheus host=example.com
ts=2017-10-11T09:36:11.450516099Z caller=http.go:150 module=https_2xx target=https://example.com/prometheus level=info msg="Received redirect" url=https://[172.20.3.94]/prometheus/
ts=2017-10-11T09:36:11.471630668Z caller=http.go:197 module=https_2xx target=https://example.com/prometheus level=info msg="Received HTTP response" status_code=404
ts=2017-10-11T09:36:11.471672664Z caller=http.go:206 module=https_2xx target=https://example.com/prometheus level=info msg="Invalid HTTP response status code" status_code=404 valid_status_codes=[200]
ts=2017-10-11T09:36:11.47171554Z caller=main.go:129 module=https_2xx target=https://example.com/prometheus level=error msg="Probe failed" duration_seconds=0.052254764
```

Why only **http** seems to be able to resolved to a name
```
ts=2017-10-11T09:34:43.991691598Z caller=http.go:150 module=https_2xx target=http://example.com/prometheus level=info msg="Received redirect" url=https://example.com/prometheus
```
and **https** seems to be able to resolved to an ip instead
```
ts=2017-10-11T09:36:11.450516099Z caller=http.go:150 module=https_2xx target=https://example.com/prometheus level=info msg="Received redirect" url=https://[172.20.3.94]/prometheus/
```

I can access it via browser (https://example.com/prometheus redirects to https://example.com/prometheus/graph) and through http too.(http force to be https)

Even curl can follow to the final (curl with http gives the same result as curl with https)
```
> curl -vL https://example.com/prometheus
*   Trying 172.20.3.43...
* Connected to example.com (172.20.3.43) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: *.thinknet.co.th
* Server certificate: COMODO RSA Domain Validation Secure Server CA
* Server certificate: COMODO RSA Certification Authority
* Server certificate: AddTrust External CA Root
> GET /prometheus HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 09:49:29 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 47
< Connection: keep-alive
< Location: /prometheus/
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
* Ignoring the response-body
* Connection #0 to host example.com left intact
* Issue another request to this URL: 'https://example.com/prometheus/'
* Found bundle for host example.com: 0x7fc53ac10020
* Re-using existing connection! (#0) with host example.com
* Connected to example.com (172.20.3.43) port 443 (#0)
> GET /prometheus/ HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 302 Found
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 09:49:29 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 40
< Connection: keep-alive
< Location: /prometheus/graph
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
* Ignoring the response-body
* Connection #0 to host example.com left intact
* Issue another request to this URL: 'https://example.com/prometheus/graph'
* Found bundle for host example.com: 0x7fc53ac10020
* Re-using existing connection! (#0) with host example.com
* Connected to example.com (172.20.3.43) port 443 (#0)
> GET /prometheus/graph HTTP/1.1
> Host: example.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.13.5
< Date: Wed, 11 Oct 2017 09:49:29 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Strict-Transport-Security: max-age=15724800; includeSubDomains;
< 
```

And more than that
The prober can probes my kubernetes dashboard and grafana too (from ingress in the same cluster with the same configuration too)
I guess that because they don't have any redirects.
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https://dashboard.example.com\&debug\=true
Logs for the probe:
ts=2017-10-11T09:55:43.278218322Z caller=main.go:116 module=https_2xx target=https://dashboard.example.com level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-11T09:55:43.278311177Z caller=utils.go:43 module=https_2xx target=https://dashboard.example.com level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2017-10-11T09:55:43.282166839Z caller=utils.go:50 module=https_2xx target=https://dashboard.example.com level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address dashboard.example.com: no suitable address found"
ts=2017-10-11T09:55:43.286418273Z caller=utils.go:63 module=https_2xx target=https://dashboard.example.com level=info msg="Resolved target address" ip=172.20.3.43
ts=2017-10-11T09:55:43.286450326Z caller=http.go:190 module=https_2xx target=https://dashboard.example.com level=info msg="Making HTTP request" url=https://[172.20.3.43] host=dashboard.example.com
ts=2017-10-11T09:55:43.311165991Z caller=http.go:197 module=https_2xx target=https://dashboard.example.com level=info msg="Received HTTP response" status_code=200
ts=2017-10-11T09:55:43.31123274Z caller=main.go:127 module=https_2xx target=https://dashboard.example.com level=info msg="Probe succeeded" duration_seconds=0.03297433
```

**Comment by brian-brazil** [comment]
This is not a problem with the blackbox exporter, your endpoint handles http and https differently.

**Comment by nutchalum** [comment]
I'll try to dig down to what's wrong with it, so this issue can be closed for now.

**Comment by nutchalum** [comment]
Somehow, I made it out

I changed my ingress
from
```
  - host: example.com
    http:
      paths:
      - path: /prometheus
        backend:
          serviceName: prometheus-federate
          servicePort: 9090
```
to
```
  - host: example.com
    http:
      paths:
      - path: /prometheus/
        backend:
          serviceName: prometheus-federate
          servicePort: 9090
```
And it worked! (adding "/" to path)

Any idea about this?

Log without "/" at the end.
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https%3A%2F%2Fexample.com%2Fprometheus\&debug\=true
Logs for the probe:
ts=2017-10-12T04:25:20.906896245Z caller=main.go:116 module=https_2xx target=https://example.com/prometheus level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-12T04:25:20.907009478Z caller=utils.go:43 module=https_2xx target=https://example.com/prometheus level=info msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-10-12T04:25:20.91105238Z caller=utils.go:63 module=https_2xx target=https://example.com/prometheus level=info msg="Resolved target address" ip=172.20.3.94
ts=2017-10-12T04:25:20.911086686Z caller=http.go:190 module=https_2xx target=https://example.com/prometheus level=info msg="Making HTTP request" url=https://[172.20.3.94]/prometheus host=example.com
ts=2017-10-12T04:25:20.936328669Z caller=http.go:150 module=https_2xx target=https://example.com/prometheus level=info msg="Received redirect" url=https://[172.20.3.94]/prometheus/
ts=2017-10-12T04:25:20.959017736Z caller=http.go:197 module=https_2xx target=https://example.com/prometheus level=info msg="Received HTTP response" status_code=404
ts=2017-10-12T04:25:20.959058989Z caller=http.go:206 module=https_2xx target=https://example.com/prometheus level=info msg="Invalid HTTP response status code" status_code=404 valid_status_codes=[200]
ts=2017-10-12T04:25:20.95910078Z caller=main.go:129 module=https_2xx target=https://example.com/prometheus level=error msg="Probe failed" duration_seconds=0.052165504
```

Log with "/" at the end
```
> curl http://localhost:9115/probe\?module\=https_2xx\&target\=https%3A%2F%2Fexample.com%2Fprometheus\&debug\=true
Logs for the probe:
ts=2017-10-12T04:25:10.884819361Z caller=main.go:116 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Beginning probe" probe=http timeout_seconds=4.5
ts=2017-10-12T04:25:10.884941926Z caller=utils.go:43 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-10-12T04:25:10.888793524Z caller=utils.go:63 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Resolved target address" ip=172.20.3.94
ts=2017-10-12T04:25:10.888828696Z caller=http.go:190 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Making HTTP request" url=https://[172.20.3.94]/prometheus host=exmaple.com
ts=2017-10-12T04:25:10.910860137Z caller=http.go:150 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Received redirect" url=https://exmaple.com/prometheus/
ts=2017-10-12T04:25:10.938948996Z caller=http.go:150 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Received redirect" url=https://exmaple.com/prometheus/graph
ts=2017-10-12T04:25:10.971166764Z caller=http.go:197 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Received HTTP response" status_code=200
ts=2017-10-12T04:25:10.971239722Z caller=main.go:127 module=https_2xx target=https://exmaple.com/prometheus level=info msg="Probe succeeded" duration_seconds=0.086361224
```

**Comment by brian-brazil** [comment]
This is still your endpoint treating http and https differently.
